clippy_lints/src/types.rs

```diff
@@ -267,6 +295,27 @@ impl Types {
                 let res = qpath_res(cx, qpath, hir_id);
                 if let Some(def_id) = res.opt_def_id() {
                     if Some(def_id) == cx.tcx.lang_items().owned_box() {
+                        if_chain! {
+                            let last = last_path_segment(qpath);
+                            if let Some(ref params) = last.args;
+                            if !params.parenthesized;
+                            if let Some(ty) = params.args.iter().find_map(|arg| match arg {
+                                GenericArg::Type(ty) => Some(ty),
+                                _ => None,
+                            });
+                            if let TyKind::Rptr(ref _lifetime_opt, ref _mut) = ty.kind;
+                            let ty_ty = hir_ty_to_ty(cx.tcx, ty);
+                            then {
+                                span_lint_and_help(
+                                    cx,
+                                    BOX_BORROW,
+                                    hir_ty.span,
+                                    "usage of `Box<&T>`",
+                                    format!("try `{}`", ty_ty).as_str(),
```

Comment:
Does the type formatting also work for user defined types, like structs, enums, ...? Can you add tests for these?

Reply:
yes it works, tests added.